discussion_grading/discussion_grading.py

```diff
@@ -44,35 +153,223 @@ def resource_string(self, path: str) -> str:
+
+        if number_of_participations >= self.number_of_participations:
+            return 1
+
+        if self.grading_method == DiscussionGradingMethod.MINIMUM_PARTICIPATIONS.name:
+            return int(number_of_participations >= self.number_of_participations)
+        elif self.grading_method == DiscussionGradingMethod.AVERAGE_PARTICIPATIONS.name:
+            return number_of_participations / self.number_of_participations
+        return 0
+
+    def get_user_stats(self) -> dict:
+        """
+        Get the user stats for the current user.
+
+        These stats include the number of:
+            * threads: learner create a post.
+            * responses: learner respond to a post.
+            * replies: learner comment on response.
+
+        If the discussion forum is not enabled or the user
+        stats are not found, a message is returned.
+
+        Example:
+        >>> self.get_user_stats()
+            {
+                "sucess": True,
+                "user_stats": {
+                    "threads": 1,
+                    "responses": 2,
+                    "replies": 3,
+                },
+            }
+
+        Returns:
+            dict: The user stats for the current user or a message.
+        """
+        try:
+            user_stats = get_course_user_stats(self.block_course_id).get("user_stats")
+        except Exception:  # pylint: disable=broad-except
+            return {
+                "sucess": False,
+                "message": _("Discussion forum is not enabled."),
+            }
+
+        for user_stat in user_stats:
+            if user_stat.get("username") == get_username(self.current_user):
+                return {
+                    "sucess": True,
+                    "user_stats": {
+                        "threads": user_stat.get("threads"),
+                        "responses": user_stat.get("responses"),
+                        "replies": user_stat.get("replies"),
+                    },
+                }
+
+        return {
+            "sucess": False,
+            "message": _("User stats not found."),
+        }
+
+    @XBlock.json_handler
+    def calculate_grade(self, _data: dict, _suffix: str = "") -> dict:
+        """
+        Calculate the grade for the student.
+
+        The grading is calculated according to the discussion participations and grading method.
+
+        Args:
+            _data (dict): Additional data to be used in the calculation.
+            _suffix (str, optional): Suffix for the handler. Defaults to "".
+
+        Returns:
+            dict: A dictionary containing the handler result.
+        """
+        if self.max_attempts and self.attempts >= self.max_attempts:
+            return {
+                "success": False,
+                "message": _("You have reached the maximum number of attempts."),
+            }
+
+        user_stats = self.get_user_stats()
+
+        if not user_stats["sucess"]:
+            return user_stats
```

Comment:
What do you think about returning the error here instead of get_user_stats(), since this is the handler itself.

Reply:
Yes, I agree
Done: https://github.com/eduNEXT/xblock-discussion-grading/commit/c8b8783012e9449883e420262590e37d5868e478